### ISSUE TITLE
Enable defaulting to `attr.NOTHING` without throwing out init

### ISSUE THREAD

**Issue by effigies** [python-attrs/attrs#764]
I have a use-case where I'm using `attr.s` to define interfaces between computational nodes, e.g.,

```Python
@promote
def f(a: int) -> int:
    return a * 2

@promote
def g(a: int, b: int) -> int:
    return a + b
```

This results in objects like:

```Python
@attr.s(kw_only=True)
class f_inputs:
    a: int = attr.ib()

@attr.s(kw_only=True)
class f_outputs:
    out: int = attr.ib()

class f:
    def __init__(self):
        self.inputs = f_inputs()
        self.outputs = f_outputs()

    @property
    def lazy_outputs(self):
        # Object that contains a future for each name in self.outputs

# ... Similar for g
```

So I might compose these like so:

```Python
f1 = f()
f2 = f(a=1)
g1 = g(a=f1.lazy_outputs.out, b=f2.lazy_outputs.out)
```

As you can see, we want to be able to instantiate without values (we might connect something up later, or a parameter may be omittable in some contexts) or provide them at instantiation.

Right now, we can either permit values to be `attr.NOTHING` by setting `@attr.s(init=False)`, which disables passing values at instantiation, or permit values which then fail on `attr.NOTHING` unless it is explicitly passed. (We also set `@attr.s(kw_only=True)`, though that is mostly incidental for this discussion.)

I can see a few options:

1) Creating our own `myattr.NOTHING` and `myattrib()` that defaults to `myattr.NOTHING` to avoid making users of our tool explicitly pass `attr.NOTHING`.
2) Have our own `initialize_class` function that is basically `f(**{field.name: field.default for field in attr.fields(f_inputs)})` followed by an update from `kwargs`
3) Add a parameter to `attr.s()` to enable the `attr.NOTHING` default without throwing out value passing

Or is there already a way to do what we want? As much as possible, we would like to avoid hacking around `attrs` so that the semantics are close to what someone familiar with `attrs` would expect.

**Comment by hynek** [comment]
I'd have to lie if I said that I fully understand what you're trying to do, but aren't defaults what you want to do?

```python
@attr.define
class f_inputs:
    a: Optional[int] = None
```

and then handle the lack of a in your code?

**Comment by effigies** [comment]
The idea is that we want to distinguish between "no input" and "a value of `None` was passed". Since `attr.NOTHING` is how attrs naturally represents no input, we'd hoped there was a way to tell attrs not to complain about this case.

I spent some time trying to think of compelling use-cases for this distinction, though, and I think perhaps we can get farther with typing and defaulting to `None` than we thought when we made this choice some time back. Going to work through some ideas for rearchitecting, and can re-open if this distinction remains worth drawing.
